sheets/api.py

```diff
@@ -330,42 +338,71 @@ def __init__(self):
+        Args:
+            enumerated_data_rows (iterable of (int, list of str)): Row indices paired with a list of strings
+                representing the data in each row
+
+        Returns:
+            (list of ProcessedRequest, list of FailedRequest, list of ProcessedRequest):
+                A list of objects representing each request row that was processed,
+                a list of objects representing each request row that couldn't be processed,
+                and a list of objects representing each request row that is complete but does't say so in
+                the spreadsheet.
         """
-        enumerated_data_rows = enumerate(
-            get_data_rows(self.coupon_request_sheet),
-            start=coupon_request_sheet_spec.first_data_row,
-        )
+        processed_requests = []
+        failed_requests = []
+        unrecorded_complete_requests = []
         for row_index, row_data in enumerated_data_rows:
             try:
-                yield row_index, CouponRequestRow.parse_raw_data(row_data)
+                coupon_req_row = CouponRequestRow.parse_raw_data(row_index, row_data)
             except Exception as exc:
                 log.error(
                     "Coupon request row could not be parsed (row %d). Exception: %s"
-                    % row_index,
-                    str(exc),
+                    % (row_index, str(exc))
+                )
+                failed_requests.append(
+                    FailedRequest(
+                        row_index=row_index,
+                        exception=exc,
+                        error_text="Row parsing error ({})".format(exc),
+                    )
                 )
                 continue
 
-    def create_coupons_from_sheet(self):
-        """
-        Creates coupons for all rows in the coupon request sheet that indicate that they are not yet processed.
-
-        Returns:
-            list of ProcessedRequest: A list of objects containing information about each request row that was processed.
-        """
-        processed_requests = []
-        for row_index, coupon_req_row in self.parsed_row_iterator():
             if coupon_req_row.date_processed:
                 continue
-            coupon_gen_request = create_coupons_for_request_row(coupon_req_row)
-            if coupon_gen_request:
+
+            try:
+                coupon_gen_request = create_coupons_for_request_row(coupon_req_row)
+            except SheetOutOfSyncException as exc:
+                unrecorded_complete_requests.append(
+                    ProcessedRequest(
+                        row_index=row_index,
+                        coupon_req_row=coupon_req_row,
+                        request_id=exc.coupon_gen_request.id,
+                        date_processed=exc.coupon_gen_request.updated_on,
+                    )
+                )
+            except Exception as exc:
+                if isinstance(exc, Company.DoesNotExist):
```

Comment:
It looks like this part is not covered by tests. Can you add tests for this? https://codecov.io/gh/mitodl/mitxpro/pull/1349/diff#D7-392